app/src/main/cpp/skyline/gpu/texture/texture.h

```diff
@@ -0,0 +1,352 @@
+    };
+
+    /**
+     * @brief A texture which is backed by host constructs while being synchronized with the underlying guest texture
+     * @note This class conforms to the Lockable and BasicLockable C++ named requirements
+     */
+    class Texture : public std::enable_shared_from_this<Texture>, public FenceCycleDependency {
+      private:
+        GPU &gpu;
+        std::mutex mutex; //!< Synchronizes any mutations to the texture or its backing
+        std::condition_variable backingCondition; //!< Signalled when a valid backing has been swapped in
+        using BackingType = std::variant<vk::Image, vk::raii::Image, memory::Image>;
+        BackingType backing; //!< The Vulkan image that backs this texture, it is nullable
+        std::shared_ptr<FenceCycle> cycle; //!< A fence cycle for when any host operation mutating the texture has completed, it must be waited on prior to any mutations to the backing
+        vk::ImageLayout layout;
+
+        /**
+         * @note The handle returned is nullable and the appropriate precautions should be taken
+         */
+        constexpr vk::Image GetBacking() {
+            return std::visit(VariantVisitor{
+                [](vk::Image image) { return image; },
+                [](const vk::raii::Image &image) { return *image; },
+                [](const memory::Image &image) { return image.vkImage; },
+            }, backing);
+        }
+
+      public:
+        std::shared_ptr<GuestTexture> guest; //!< The guest texture from which this was created, it's required for syncing
+        texture::Dimensions dimensions;
+        texture::Format format;
+        vk::ImageTiling tiling;
+        vk::ComponentMapping mapping;
+
+        Texture(GPU &gpu, BackingType &&backing, std::shared_ptr<GuestTexture> guest, texture::Dimensions dimensions, const texture::Format &format, vk::ImageLayout layout, vk::ImageTiling tiling, vk::ComponentMapping mapping);
+
+        Texture(GPU &gpu, BackingType &&backing, texture::Dimensions dimensions, const texture::Format &format, vk::ImageLayout layout, vk::ImageTiling tiling, vk::ComponentMapping mapping = {});
+
+        /**
+         * @brief Creates and allocates memory for the backing to creates a texture object wrapping it
+         * @param usage Usage flags that will applied aside from VK_IMAGE_USAGE_TRANSFER_SRC_BIT/VK_IMAGE_USAGE_TRANSFER_DST_BIT which are mandatory
+         */
+        Texture(GPU &gpu, texture::Dimensions dimensions, const texture::Format &format, vk::ImageLayout initialLayout = vk::ImageLayout::eGeneral, vk::ImageUsageFlags usage = {}, vk::ImageTiling tiling = vk::ImageTiling::eOptimal, vk::ComponentMapping mapping = {});
+
+        /**
+         * @brief Acquires an exclusive lock on the texture for the calling thread
```

Comment:
Note that these are lower case so they can be used with unique_lock

Reply:
See the class comment `@note This class conforms to the Lockable and BasicLockable C++ named requirements`